language-plutus-core/src/Language/PlutusCore/Renamer.hs

```diff
@@ -112,75 +117,143 @@ annotateTy (TyApp x ty ty') =
 annotateTy (TyInt x n) = pure (TyInt x n)
 
--- | Mapping from locally unique indices to globally unique 'Unique's.
-newtype UniquesRenaming = UniquesRenaming
-    { unUniquesRenaming :: IM.IntMap Unique
+-- | Mapping from locally unique indices to globally unique uniques.
+newtype UniquesRenaming unique = UniquesRenaming
+    { unUniquesRenaming :: IM.IntMap unique
     }
 
--- | The monad the renamer runs in.
-type RenameM = ReaderT UniquesRenaming Quote
+-- | Scoping-aware mapping from locally unique indices to globally unique uniques.
+data ScopedUniquesRenaming = ScopedUniquesRenaming
+    { _scopedUniquesRenamingTypes :: UniquesRenaming TypeUnique
+    , _scopedUniquesRenamingTerms :: UniquesRenaming TermUnique
+    }
 
--- | Rename 'Unique's so that they're globally unique.
-rename :: (Rename a, MonadQuote m) => a -> m a
-rename x = liftQuote $ runReaderT (renameM x) (UniquesRenaming mempty)
+makeLenses ''ScopedUniquesRenaming
 
--- | Save the mapping from an old 'Unique' to a new one.
-updateUniquesRenaming :: Unique -> Unique -> UniquesRenaming -> UniquesRenaming
-updateUniquesRenaming uniqOld uniqNew (UniquesRenaming uniques) =
-    UniquesRenaming $ IM.insert (unUnique uniqOld) uniqNew uniques
+-- | A class that specifies which 'UniquesRenaming' a @renaming@ has inside.
+-- A @renaming@ can contain several 'UniquesRenaming's (like 'ScopedUniquesRenaming', for example).
+class Coercible unique Unique => HasUniquesRenaming renaming unique where
+    uniquesRenaming :: Lens' renaming (UniquesRenaming unique)
 
--- | Look up a new 'Unique' an old 'Unique' got mapped to.
-lookupUnique :: Unique -> RenameM (Maybe Unique)
-lookupUnique uniq = asks $ IM.lookup (unUnique uniq) . unUniquesRenaming
+instance (Coercible unique1 Unique, unique1 ~ unique2) =>
+        HasUniquesRenaming (UniquesRenaming unique1) unique2 where
+    uniquesRenaming = id
 
--- | Replace the 'Unique' in a value by a new 'Unique', save the mapping
--- from an old 'Unique' to the new one and supply the updated value to a continuation.
-withRefreshed :: HasUnique a => a -> (a -> RenameM c) -> RenameM c
-withRefreshed x k = do
-    let uniqOld = x ^. unique
-    uniqNew <- liftQuote freshUnique
-    local (updateUniquesRenaming uniqOld uniqNew) $ k (x & unique .~ uniqNew)
+instance HasUniquesRenaming ScopedUniquesRenaming TypeUnique where
+    uniquesRenaming = scopedUniquesRenamingTypes
+
+instance HasUniquesRenaming ScopedUniquesRenaming TermUnique where
+    uniquesRenaming = scopedUniquesRenamingTerms
 
 -- | The class of things that can be renamed.
 -- I.e. things that are capable of satisfying the global uniqueness condition.
 class Rename a where
-    renameM :: a -> RenameM a
+    -- | Rename 'Unique's so that they're globally unique.
+    -- In case there are any free variables, they must be leaved untouched.
+    -- Must always assign new names to bound variables,
+    -- so that @rename@ can be used for cloning as well.
```

Comment:
I'd prefer to say "alpha renaming" instead of "cloning", since that's what it is and it's an established name.

Reply:
Yeah, much clearer.